File: IAIso-v5.0/core/iaiso-java/iaiso-audit/src/main/java/io/iaiso/audit/JsonlFileSink.java
package io.iaiso.audit;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.concurrent.locks.ReentrantLock;

/**
 * A sink that appends one JSON event per line to a file.
 * <p>
 * Errors are silently dropped — best-effort delivery semantics
 * specified in {@code spec/events/README.md §6}.
 */
public final class JsonlFileSink implements Sink {
    private final Path path;
    private final ReentrantLock lock = new ReentrantLock();

    public JsonlFileSink(Path path) {
        this.path = path;
    }

    @Override
    public void emit(Event event) {
        lock.lock();
        try {
            String line = event.toJson() + "\n";
            Files.writeString(path, line, StandardCharsets.UTF_8,
                StandardOpenOption.CREATE, StandardOpenOption.APPEND);
        } catch (IOException ignored) {
            // best-effort
        } finally {
            lock.unlock();
        }
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Audit/JsonlFileSink.php
<?php

declare(strict_types=1);

namespace IAIso\Audit;

/** Appends events as JSONL to a file. I/O errors are silently dropped. */
final class JsonlFileSink implements Sink
{
    public function __construct(private readonly string $path)
    {
    }

    public function emit(Event $event): void
    {
        $line = $event->toJson() . PHP_EOL;
        // LOCK_EX to keep multi-process writers from interleaving lines.
        @file_put_contents($this->path, $line, FILE_APPEND | LOCK_EX);
    }
}
